stackslib/src/net/mod.rs

```diff
@@ -1563,6 +1562,539 @@ impl NetworkResult {
         }
     }
 
+    /// Get the set of all StacksBlocks represented
+    fn all_block_ids(&self) -> HashSet<StacksBlockId> {
+        let mut blocks: HashSet<_> = self
+            .blocks
+            .iter()
+            .map(|(ch, blk, _)| StacksBlockId::new(&ch, &blk.block_hash()))
+            .collect();
+
+        let pushed_blocks: HashSet<_> = self
+            .pushed_blocks
+            .iter()
+            .map(|(_, block_list)| {
+                block_list
+                    .iter()
+                    .map(|block_data| {
+                        block_data
+                            .blocks
+                            .iter()
+                            .map(|block_datum| {
+                                StacksBlockId::new(&block_datum.0, &block_datum.1.block_hash())
+                            })
+                            .collect::<HashSet<_>>()
+                    })
+                    .collect::<Vec<HashSet<_>>>()
+            })
+            .collect::<Vec<Vec<HashSet<_>>>>()
+            .into_iter()
+            .flatten()
+            .into_iter()
+            .fold(HashSet::new(), |mut acc, next| {
+                acc.extend(next.into_iter());
+                acc
+            });
```

Comment:
You can flatten while they're still iterators:

```suggestion
            .map(|(_, block_list)| {
                block_list
                    .iter()
                    .map(|block_data| {
                        block_data
                            .blocks
                            .iter()
                            .map(|block_datum| {
                                StacksBlockId::new(&block_datum.0, &block_datum.1.block_hash())
                            })
                    })
                    .flatten()
            })
            .flatten()
            .collect();
```